### PR TITLE
Properly resolve method references in constructed generic types.

### LINKED ISSUE #53802: Address the PROTOTYPE comment in TypeSymbol.ReportImplicitImplementationMatchDiagnostics
```
            // In constructed types, it is possible to see multiple members with the same (runtime) signature.
            // Now that we know which member will implement the interface member, confirm that it is the only
            // such member.
            if (!implicitImpl.ContainingType.IsDefinition)
            {
                foreach (Symbol member in implicitImpl.ContainingType.GetMembers(implicitImpl.Name))
                {
                    if (member.DeclaredAccessibility != Accessibility.Public || member.IsStatic || member == implicitImpl)
                    {
                        //do nothing - not an ambiguous implementation

                        // PROTOTYPE(StaticAbstractMembersInInterfaces): We likely need to do something here for static members.
                    }
                    else if (MemberSignatureComparer.RuntimeImplicitImplementationComparer.Equals(interfaceMember, member) && !member.IsAccessor())
                    {
                        // CONSIDER: Dev10 does not seem to report this for indexers or their accessors.
                        diagnostics.Add(ErrorCode.WRN_MultipleRuntimeImplementationMatches, GetImplicitImplementationDiagnosticLocation(interfaceMember, implementingType, member), member, interfaceMember, implementingType);
                    }
                }
            }
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Also closes #53802.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs [hunk 1]
```diff
@@ -131,7 +131,7 @@ private void CheckAbstractClassImplementations(BindingDiagnosticBag diagnostics)
 
                 HasBaseTypeDeclaringInterfaceResult? hasBaseClassDeclaringInterface = null;
 
-                foreach (var interfaceMember in @interface.GetMembersUnordered())
+                foreach (var interfaceMember in @interface.GetMembers())
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:134]: Could you please explain the impact of this change?

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 2]
```diff
@@ -25088,5 +25088,263 @@ class Test
...
+
+                Assert.Equal("void Base<T>.Method(T i)", bM2.ToTestDisplayString());
+                Assert.Same(bM2, b.FindImplementationForInterfaceMember(biMethods[2]));
+
+                var bM1Impl = ((MethodSymbol)bM1).ExplicitInterfaceImplementations;
+                var bM2Impl = ((MethodSymbol)bM2).ExplicitInterfaceImplementations;
+
+                if (module is PEModuleSymbol)
+                {
+                    Assert.Equal(biMethods[0], bM1Impl.Single());
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:25150]: Should this be `Assert.Same`? (also for asserts on `biMethods[1]`, `biMethods[2]` below)

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:134]: > Could you please explain the impact of this change?

We collect and emit MethodImpl entries in deterministic order.

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:25150]: > Should this be Assert.Same? (also for asserts on biMethods[1], biMethods[2] below)

When we are dealing with generic instantiations, Assert.Same doesn't work. The symbols are not unique, but often equal and represent the same entity.

**Pull request merged**